File: ml-benchmark/autolifter/single-pass/mpp.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = -3;
Config SampleIntMax = 3;
Config NonLinear = true;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

/*User provided programs*/

max = \a: Int. \b: Int. if < a b then b else a;

mpp = fix (lambda f: Int -> List -> Int. \pre: Int. lambda l: List.
  match l with
    nil _ -> pre
  | cons {h, t} ->
    max pre (f (* h pre) t)
  end
) 1;

main = single_pass mpp;

File: ml-benchmark/autolifter/single-pass/mpp.ml
type lst = Nil | Cons of int * lst

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let max a b = if a < b then b else a

let rec mpp_f pre l =
  match l with
  | Nil -> pre
  | Cons (h, t) -> max pre (mpp_f (h * pre) t)

let mpp l = mpp_f 1 l

let program = single_pass mpp